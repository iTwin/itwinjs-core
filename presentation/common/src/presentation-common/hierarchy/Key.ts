/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Hierarchies
 */

import { InstanceKey, InstanceKeyJSON } from "../EC";

/**
 * Standard node types
 * @public
 */
export enum StandardNodeTypes {
  ECInstancesNode = "ECInstancesNode",
  ECClassGroupingNode = "ECClassGroupingNode",
  ECPropertyGroupingNode = "ECPropertyGroupingNode",
  DisplayLabelGroupingNode = "DisplayLabelGroupingNode",
}

/**
 * One of the node key types
 * @public
 */
export type NodeKey = BaseNodeKey | ECInstancesNodeKey | ECClassGroupingNodeKey | ECPropertyGroupingNodeKey | LabelGroupingNodeKey;
/** @public */
export namespace NodeKey { // eslint-disable-line @typescript-eslint/no-redeclare
  /** Serialize given [[NodeKey]] to JSON */
  export function toJSON(key: NodeKey): NodeKeyJSON {
    if (isInstancesNodeKey(key))
      return { ...key, instanceKeys: key.instanceKeys.map(InstanceKey.toJSON) };
    return { ...key };
  }

  /** Deserialize node key from JSON */
  export function fromJSON(json: NodeKeyJSON): NodeKey {
    if (isInstancesNodeKey(json))
      return { ...json, instanceKeys: json.instanceKeys.map(InstanceKey.fromJSON) };
    return { ...json };
  }

  /** Checks if the supplied key is an [[ECInstancesNodeKey]] */
  export function isInstancesNodeKey(key: NodeKey): key is ECInstancesNodeKey {
    return key.type === StandardNodeTypes.ECInstancesNode;
  }

  /** Checks if the supplied key is an [[ECClassGroupingNodeKey]] */
  export function isClassGroupingNodeKey(key: NodeKey): key is ECClassGroupingNodeKey {
    return key.type === StandardNodeTypes.ECClassGroupingNode;
  }

  /** Checks if the supplied key is an [[ECPropertyGroupingNodeKey]] */
  export function isPropertyGroupingNodeKey(key: NodeKey): key is ECPropertyGroupingNodeKey {
    return key.type === StandardNodeTypes.ECPropertyGroupingNode;
  }

  /** Checks if the supplied key is a [[LabelGroupingNodeKey]] */
  export function isLabelGroupingNodeKey(key: NodeKey): key is LabelGroupingNodeKey {
    return key.type === StandardNodeTypes.DisplayLabelGroupingNode;
  }

  /** Checks if the supplied key is a grouping node key */
  export function isGroupingNodeKey(key: NodeKey): key is GroupingNodeKey {
    return isClassGroupingNodeKey(key) || isPropertyGroupingNodeKey(key) || isLabelGroupingNodeKey(key);
  }
}

/**
 * Node key path. Can be used to define path from one node to another.
 * @public
 */
export type NodeKeyPath = NodeKey[];

/**
 * Data structure that describes a basic node key
 * @public
 */
export interface BaseNodeKey {
  /** Node type */
  type: string;
  /** Node hash path from root to the node whose key this is */
  pathFromRoot: string[];
}

/**
 * Data structure that describes a node ECInstance node key
 * @public
 */
export interface ECInstancesNodeKey extends BaseNodeKey {
  type: StandardNodeTypes.ECInstancesNode;
  /** List of [[InstanceKey]] objects of ECInstances represented by the node */
  instanceKeys: InstanceKey[];
}

/**
 * Serialized [[ECInstancesNodeKey]] JSON representation.
 * @public
 */
export interface ECInstancesNodeKeyJSON extends BaseNodeKey {
  type: StandardNodeTypes.ECInstancesNode;
  instanceKeys: InstanceKeyJSON[];
}

/**
 * Data structure that describes a grouping node key
 * @public
 */
export interface GroupingNodeKey extends BaseNodeKey {
  /**
   * Get the number of instances grouped by the node represented
   * by this key.
   *
   * **Note:** this property is just a helper and is not involved
   * in identifying a node.
   */
  groupedInstancesCount: number;
}

/**
 * Data structure that describes an ECClass grouping node key
 * @public
 */
export interface ECClassGroupingNodeKey extends GroupingNodeKey {
  type: StandardNodeTypes.ECClassGroupingNode;
  /** Full name of the grouping ECClass */
  className: string;
}

/**
 * Data structure that describes an ECProperty grouping node key
 * @public
 */
export interface ECPropertyGroupingNodeKey extends GroupingNodeKey {
  type: StandardNodeTypes.ECPropertyGroupingNode;
  /** Full name of the grouping ECProperty class */
  className: string;
  /** Name of the ECProperty */
  propertyName: string;
  /**
   * Grouping value
   * @deprecated Use [[groupingValues]] - one node might be grouping by multiple raw values.
   */
  groupingValue: any;
  /** Raw grouping values */
  groupingValues: any[];
}

/**
 * Data structure that describes a display label grouping node key
 * @public
 */
export interface LabelGroupingNodeKey extends GroupingNodeKey {
  type: StandardNodeTypes.DisplayLabelGroupingNode;
  /** Grouping display label */
  label: string;
}

/**
 * One of the serialized node key types
 * @public
 */
export type NodeKeyJSON = BaseNodeKey | ECInstancesNodeKeyJSON | ECClassGroupingNodeKey | ECPropertyGroupingNodeKey | LabelGroupingNodeKey;
