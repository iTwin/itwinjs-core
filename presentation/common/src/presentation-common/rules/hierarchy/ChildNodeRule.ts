/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module PresentationRules
 */

import { NavigationRuleBase } from "./NavigationRule";

/**
 * Child node rules are used to define child nodes in a hierarchy.
 *
 * @see [Child node rule reference documentation page]($docs/presentation/hierarchies/ChildNodeRule.md)
 * @public
 */
export interface ChildNodeRule extends NavigationRuleBase {
  /** Used for serializing to JSON. */
  ruleType: "ChildNodes";
}
