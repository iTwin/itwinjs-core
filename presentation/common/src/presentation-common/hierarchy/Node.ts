/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Hierarchies
 */

import { LabelDefinition, LabelDefinitionJSON } from "../LabelDefinition";
import { NodeKey, NodeKeyJSON } from "./Key";

/**
 * Data structure that describes a tree node.
 * @public
 */
export interface Node {
  /** A key that uniquely identifies a node. */
  key: NodeKey;
  /** Definition of node display label */
  label: LabelDefinition;
  /** Extensive description */
  description?: string;
  /** Image ID */
  imageId?: string;
  /** Foreground color */
  foreColor?: string;
  /** Background color */
  backColor?: string;
  /** Font style */
  fontStyle?: string;
  /** Does this node have child nodes */
  hasChildren?: boolean;
  /** Is this node selectable */
  isSelectionDisabled?: boolean;
  /** Is this node editable */
  isEditable?: boolean;
  /** Is this node expanded */
  isExpanded?: boolean;
  /** Is checkbox visible for this node */
  isCheckboxVisible?: boolean;
  /** Is this node's checkbox checked */
  isChecked?: boolean;
  /** Is this node's checkbox enabled */
  isCheckboxEnabled?: boolean;
  /** Extended data injected into this node */
  extendedData?: {
    [key: string]: any;
  };
}

/**
 * Serialized [[Node]] JSON representation.
 * @public
 */
export interface NodeJSON {
  key: NodeKeyJSON;
  labelDefinition: LabelDefinitionJSON;
  description?: string;
  imageId?: string;
  foreColor?: string;
  backColor?: string;
  fontStyle?: string;
  hasChildren?: boolean;
  isSelectionDisabled?: boolean;
  isEditable?: boolean;
  isExpanded?: boolean;
  isCheckboxVisible?: boolean;
  isChecked?: boolean;
  isCheckboxEnabled?: boolean;
  extendedData?: {
    [key: string]: any;
  };
}

/** @alpha */
export type PartialNode = AllOrNone<Partial<Node>, "key" | "label">;
/** @alpha */
export type PartialNodeJSON = AllOrNone<Partial<NodeJSON>, "key" | "labelDefinition">;
type AllOrNone<T, P extends keyof T> = Omit<T, P> & ({ [K in P]?: never } | Required<Pick<T, P>>);

/** @public */
export namespace Node {
  /** Serialize given [[Node]] to JSON */
  export function toJSON(node: Node): NodeJSON {
    const { key, label, ...baseNode } = node;
    return {
      ...baseNode,
      key: NodeKey.toJSON(key),
      labelDefinition: LabelDefinition.toJSON(label),
    };
  }

  /** @internal */
  export function toPartialJSON(node: PartialNode): PartialNodeJSON {
    if (node.key === undefined) {
      return node;
    }

    const { key, label, ...baseNode } = node;
    return {
      ...baseNode,
      key: NodeKey.toJSON(key),
      labelDefinition: LabelDefinition.toJSON(label),
    };
  }

  /** Deserialize [[Node]] from JSON */
  export function fromJSON(json: NodeJSON | string): Node {
    if (typeof json === "string")
      return JSON.parse(json, reviver);
    const { labelDefinition, ...baseJson } = json;
    return {
      ...baseJson,
      key: NodeKey.fromJSON(json.key),
      label: LabelDefinition.fromJSON(labelDefinition),
    };
  }

  /** @internal */
  export function fromPartialJSON(json: PartialNodeJSON): PartialNode {
    if (json.key === undefined) {
      return json;
    }

    const { key, labelDefinition, ...baseJson } = json;
    return {
      ...baseJson,
      key: NodeKey.fromJSON(key),
      label: LabelDefinition.fromJSON(labelDefinition),
    };
  }

  /**
   * Reviver function that can be used as a second argument for
   * `JSON.parse` method when parsing [[Node]] objects.
   *
   * @internal
   */
  export function reviver(key: string, value: any): any {
    return key === "" ? fromJSON(value) : value;
  }

  /**
   * Deserialize nodes list from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized nodes list
   *
   * @internal
   */
  export function listFromJSON(json: NodeJSON[] | string): Node[] {
    if (typeof json === "string")
      return JSON.parse(json, listReviver);
    return json.map((m) => fromJSON(m));
  }

  /**
   * Reviver function that can be used as a second argument for
   * `JSON.parse` method when parsing [[Node]][] objects.
   *
   * @internal
   */
  export function listReviver(key: string, value: any): any {
    return key === "" ? listFromJSON(value) : value;
  }
}
