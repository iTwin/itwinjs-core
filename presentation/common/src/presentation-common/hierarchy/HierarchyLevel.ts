/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Hierarchies
 */

import { Node, NodeJSON } from "./Node";

/**
 * Defines a hierarchy level that consists of an array of nodes and possibly other information.
 * @public
 */
export interface HierarchyLevel {
  /** A list of nodes in a hierarchy level. */
  nodes: Node[];
  /**
   * Identifies whether the hierarchy level supports filtering. If not, requesting either a hierarchy level descriptor or
   * a hierarchy level with [[HierarchyRequestOptions.instanceFilter]] will throw an error with [[PresentationStatus.InvalidArgument]] status.
   */
  supportsFiltering?: boolean;
}

/**
 * JSON representation of [[HierarchyLevel]].
 * @beta
 * @deprecated in 3.x. Use [[HierarchyLevel]]
 */
export interface HierarchyLevelJSON {
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  nodes: NodeJSON[];
  supportsFiltering?: boolean;
}

/**
 * Contains helper functions for working with objects of [[HierarchyLevel]] type.
 * @public
 */
export namespace HierarchyLevel {
  /**
   * Deserialize [[HierarchyLevel]] from JSON
   * @beta
   * @deprecated in 3.x. Use [[HierarchyLevel]].
   */
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  export function fromJSON(json: HierarchyLevelJSON): HierarchyLevel {
    return {
      ...json,
      // eslint-disable-next-line @typescript-eslint/no-deprecated
      nodes: json.nodes.map(Node.fromJSON),
    };
  }
}
