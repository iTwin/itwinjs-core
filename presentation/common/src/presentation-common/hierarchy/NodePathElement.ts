/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Hierarchies
 */

import { Node, NodeJSON } from "./Node";

/**
 * Serialized [[NodePathElement]] JSON representation.
 * @public
 * @deprecated in 3.x. Use [[NodePathElement]].
 */
export interface NodePathElementJSON {
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  node: NodeJSON;
  index: number;
  isMarked?: boolean;
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  children: NodePathElementJSON[];
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  filteringData?: NodePathFilteringDataJSON;
}

/**
 * Describes a single step in the nodes path.
 * @public
 */
export interface NodePathElement {
  /** Node instance */
  node: Node;
  /** Node index  */
  index: number;
  /** Is this element part of the marked path */
  isMarked?: boolean;
  /** Child path elements */
  children: NodePathElement[];
  /** Additional filtering-related information */
  filteringData?: NodePathFilteringData;
}

/** @public */
export namespace NodePathElement {
  /**
   * Serialize given [[NodePathElement]] to JSON
   * @deprecated in 3.x. Use [[NodePathElement]].
   */
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  export function toJSON(npe: NodePathElement): NodePathElementJSON {
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    const result: NodePathElementJSON = {
      // eslint-disable-next-line @typescript-eslint/no-deprecated
      node: Node.toJSON(npe.node),
      index: npe.index,
      // eslint-disable-next-line @typescript-eslint/no-deprecated
      children: npe.children.map(NodePathElement.toJSON),
    };
    if (undefined !== npe.isMarked) {
      result.isMarked = npe.isMarked;
    }
    if (undefined !== npe.filteringData) {
      // eslint-disable-next-line @typescript-eslint/no-deprecated
      result.filteringData = NodePathFilteringData.toJSON(npe.filteringData);
    }
    return result;
  }

  /**
   * Deserialize [[NodePathElement]] from JSON
   * @deprecated in 3.x. Use [[NodePathElement]].
   */
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  export function fromJSON(json: NodePathElementJSON | string): NodePathElement {
    if (typeof json === "string") {
      return JSON.parse(json, reviver);
    }
    const result: NodePathElement = {
      index: json.index,
      // eslint-disable-next-line @typescript-eslint/no-deprecated
      node: Node.fromJSON(json.node),
      children: listFromJSON(json.children),
    };
    if (undefined !== json.isMarked) {
      result.isMarked = json.isMarked;
    }
    if (undefined !== json.filteringData) {
      // eslint-disable-next-line @typescript-eslint/no-deprecated
      result.filteringData = NodePathFilteringData.fromJSON(json.filteringData);
    }
    return result;
  }

  /**
   * Reviver function that can be used as a second argument for
   * `JSON.parse` method when parsing [[NodePathElement]] objects.
   * @internal
   */
  export function reviver(key: string, value: any): any {
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    return key === "" ? fromJSON(value) : value;
  }

  /**
   * Deserialize [[NodePathElement]] list from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized [[NodePathElement]] list
   * @internal
   */
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  export function listFromJSON(json: NodePathElementJSON[] | string): NodePathElement[] {
    if (typeof json === "string") {
      return JSON.parse(json, listReviver);
    }
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    return json.map((m) => fromJSON(m));
  }

  /**
   * Reviver function that can be used as a second argument for
   * `JSON.parse` method when parsing [[NodePathElement]][] objects.
   * @internal
   */
  export function listReviver(key: string, value: any): any {
    return key === "" ? listFromJSON(value) : value;
  }
}

/**
 * Serialized [[NodePathFilteringData]] JSON representation.
 * @public
 * @deprecated in 3.x. Use [[NodePathFilteringData]].
 */
export interface NodePathFilteringDataJSON {
  occurances: number;
  childrenOccurances: number;
}

/**
 * Data related to node hierarchy filtering
 * @public
 */
export interface NodePathFilteringData {
  /** Number of filter matches in the current element */
  matchesCount: number;
  /** Number of filter matches in the current element's children (recursively) */
  childMatchesCount: number;
}

/** @public */
export namespace NodePathFilteringData {
  /**
   * Serialize given [[NodePathFilteringData]] to JSON
   * @deprecated in 3.x. Use [[NodePathFilteringData]].
   */
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  export function toJSON(npfd: NodePathFilteringData): NodePathFilteringDataJSON {
    return {
      occurances: npfd.matchesCount,
      childrenOccurances: npfd.childMatchesCount,
    };
  }

  /**
   * Deserialize [[NodePathFilteringData]] from JSON
   * @deprecated in 3.x. Use [[NodePathFilteringData]].
   */
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  export function fromJSON(json: NodePathFilteringDataJSON): NodePathFilteringData {
    return {
      matchesCount: json.occurances,
      childMatchesCount: json.childrenOccurances,
    };
  }
}
