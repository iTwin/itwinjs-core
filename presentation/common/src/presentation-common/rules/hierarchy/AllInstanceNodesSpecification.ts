/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module PresentationRules
 */
import { SchemasSpecification } from "../SchemasSpecification";
import { ChildNodeSpecificationBase, ChildNodeSpecificationTypes, DefaultGroupingPropertiesContainer } from "./ChildNodeSpecification";

/**
 * Creates nodes for all available instances filtered only by the
 * [[supportedSchemas]] properties of the specification or the ruleset.
 *
 * @public
 * @deprecated Use [[InstanceNodesOfSpecificClassesSpecification]]. Will be removed in iModel.js 3.0
 */
export interface DEPRECATED_AllInstanceNodesSpecification extends ChildNodeSpecificationBase, DefaultGroupingPropertiesContainer { // eslint-disable-line @typescript-eslint/naming-convention
  /** Used for serializing to JSON. */
  specType: ChildNodeSpecificationTypes.DEPRECATED_AllInstanceNodes;

  /**
   * Specification of schemas whose instances should be returned.
   */
  supportedSchemas?: SchemasSpecification;
}
