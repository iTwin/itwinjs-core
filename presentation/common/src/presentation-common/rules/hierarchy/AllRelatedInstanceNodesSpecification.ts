/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module PresentationRules
 */

import { RelationshipDirection } from "../RelationshipDirection";
import { SchemasSpecification } from "../SchemasSpecification";
import { ChildNodeSpecificationBase, ChildNodeSpecificationTypes, DefaultGroupingPropertiesContainer } from "./ChildNodeSpecification";

/**
 * Creates nodes for all related instances for parent ECInstance node.
 *
 * **Precondition:** can be used only if parent node is ECInstance node. If there is
 * no immediate parent instance node it will go up until it finds one.
 *
 * @public
 * @deprecated Use [[RelatedInstanceNodesSpecification]]. Will be removed in iModel.js 3.0
 */
export interface DEPRECATED_AllRelatedInstanceNodesSpecification extends ChildNodeSpecificationBase, DefaultGroupingPropertiesContainer { // eslint-disable-line @typescript-eslint/naming-convention
  /** Used for serializing to JSON. */
  specType: ChildNodeSpecificationTypes.DEPRECATED_AllRelatedInstanceNodes;

  /**
   * Specification of schemas whose instances should be returned.
   */
  supportedSchemas?: SchemasSpecification;

  /**
   * Direction that will be followed in the relationship select criteria.
   * Defaults to [[RelationshipDirection.Both]].
   */
  requiredDirection?: RelationshipDirection;

  /**
   * Skips defined level of related items and shows next level related items.
   *
   * **Example:** lets say we have relationship hierarchy A->B->C, but we want
   * to show A->C. In this case we need to set this property to 1 - the engine will
   * query all direct relationships and all nested relationships and only nested
   * ones will be used to create content.
   *
   * Defaults to `0`.
   *
   * @type integer
   * @minimum 0
   */
  skipRelatedLevel?: number;
}
