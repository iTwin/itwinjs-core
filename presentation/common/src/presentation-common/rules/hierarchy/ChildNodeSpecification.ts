/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module PresentationRules
 */

import { RelatedInstanceSpecification } from "../RelatedInstanceSpecification";
import { DEPRECATED_AllInstanceNodesSpecification } from "./AllInstanceNodesSpecification";
import { DEPRECATED_AllRelatedInstanceNodesSpecification } from "./AllRelatedInstanceNodesSpecification";
import { ChildNodeRule } from "./ChildNodeRule";
import { CustomNodeSpecification } from "./CustomNodeSpecification";
import { CustomQueryInstanceNodesSpecification } from "./CustomQueryInstanceNodesSpecification";
import { InstanceNodesOfSpecificClassesSpecification } from "./InstanceNodesOfSpecificClassesSpecification";
import { RelatedInstanceNodesSpecification } from "./RelatedInstanceNodesSpecification";

/**
 * Used for serializing array of [[ChildNodeSpecification]]
 * @public
 */
export enum ChildNodeSpecificationTypes {
  // hierarchy specifications
  DEPRECATED_AllInstanceNodes = "AllInstanceNodes", // eslint-disable-line @typescript-eslint/naming-convention
  DEPRECATED_AllRelatedInstanceNodes = "AllRelatedInstanceNodes", // eslint-disable-line @typescript-eslint/naming-convention
  RelatedInstanceNodes = "RelatedInstanceNodes",
  InstanceNodesOfSpecificClasses = "InstanceNodesOfSpecificClasses",
  CustomQueryInstanceNodes = "CustomQueryInstanceNodes",
  CustomNode = "CustomNode",
}

/**
 * Base interface for all [[ChildNodeSpecification]] implementations. Not
 * meant to be used directly, see `ChildNodeSpecification`.
 *
 * @public
 */
export interface ChildNodeSpecificationBase {
  /** Used for serializing to JSON. */
  specType: ChildNodeSpecificationTypes;

  /**
   * Defines the order in which specifications are evaluated and executed. Defaults to `1000`.
   *
   * @type integer
   */
  priority?: number;

  /**
   * This tells the rules engine that nodes produced using this
   * specification always or never have children. Defaults to `Unknown`.
   *
   * **Note:** setting this flag to `Always` or `Never` improves performance.
   */
  hasChildren?: "Always" | "Never" | "Unknown";

  /**
   * Hide instance nodes provided by this specification and directly show nodes of its children.
   * This helps if you need to define related instance nodes of particular parent node that is not available in the
   * hierarchy.
   */
  hideNodesInHierarchy?: boolean;

  /**
   * Hide nodes if they don't have children.
   */
  hideIfNoChildren?: boolean;

  /**
   * An [ECExpression]($docs/learning/presentation/Hierarchies/ECExpressions.md#specification) which
   * indicates whether a node should be hidden or not.
   *
   * @note While the attribute provides much flexibility, it also has performance implications - it's
   * strongly suggested to first consider using `instanceFilter`, `hideNodesInHierarchy` or `hideIfNoChildren`
   * and only use `hideExpression` if none of them are sufficient.
   */
  hideExpression?: string;

  /**
   * Set this flag to `true` to suppress default sorting of ECInstances returned by this specification.
   *
   * **Note:** setting this flag to `true` improves performance.
   */
  doNotSort?: boolean;

  /**
   * Suppress similar ancestor nodes' checking when creating nodes based on this specification.
   *
   * By default we stop creating hierarchy at the node that has a similar ancestor representing
   * the same ECInstance and created using the same specification to prevent creating infinite
   * hierarchies. With this flag enabled, we allow up to 10 similar ancestors before we consider
   * the hierarchy infinite.
   *
   * @beta
   */
  suppressSimilarAncestorsCheck?: boolean;

  /** Specifications of related instances that can be used in nodes' creation. */
  relatedInstances?: RelatedInstanceSpecification[];

  /** [Nested rule]($docs/learning/presentation/Hierarchies/Terminology.md#nested-rules) specifications. */
  nestedRules?: ChildNodeRule[];
}

/**
 * A container of default grouping properties. Used for specifications that support
 * default grouping. Not meant to be used directly, see `ChildNodeSpecification`.
 *
 * @public
 */
export interface DefaultGroupingPropertiesContainer {
  /** Group instances by ECClass. Defaults to `true`. */
  groupByClass?: boolean;

  /** Group instances by label. Defaults to `true`. */
  groupByLabel?: boolean;
}

/**
 * Navigation rule specifications that define what content the rule results in.
 * @public
 */
export type ChildNodeSpecification =
  DEPRECATED_AllInstanceNodesSpecification | // eslint-disable-line deprecation/deprecation
  DEPRECATED_AllRelatedInstanceNodesSpecification | // eslint-disable-line deprecation/deprecation
  CustomNodeSpecification | InstanceNodesOfSpecificClassesSpecification |
  RelatedInstanceNodesSpecification | CustomQueryInstanceNodesSpecification;
