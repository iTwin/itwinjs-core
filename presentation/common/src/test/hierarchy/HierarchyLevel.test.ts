/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
import { expect } from "chai";
import { HierarchyLevel } from "../../presentation-common/hierarchy/HierarchyLevel";
import { Node } from "../../presentation-common/hierarchy/Node";
import { createTestNode } from "../_helpers/Hierarchy";

describe("HierarchyLevel", () => {
  describe("fromJSON", () => {
    it("creates valid HierarchyLevel from JSON", () => {
      // eslint-disable-next-line @typescript-eslint/no-deprecated
      const hl = HierarchyLevel.fromJSON({
        // eslint-disable-next-line @typescript-eslint/no-deprecated
        nodes: [Node.toJSON(createTestNode())],
        supportsFiltering: true,
      });
      expect(hl).to.matchSnapshot();
    });
  });
});
