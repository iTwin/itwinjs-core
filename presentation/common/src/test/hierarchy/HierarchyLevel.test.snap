// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HierarchyLevel fromJSON creates valid HierarchyLevel from JSON 1`] = `
Object {
  "nodes": Array [
    Object {
      "key": Object {
        "pathFromRoot": Array [],
        "type": "test-node",
        "version": 0,
      },
      "label": Object {
        "displayValue": "test label",
        "rawValue": "test label",
        "typeName": "string",
      },
    },
  ],
  "supportsFiltering": true,
}
`;
